nemo_aligner/algorithms/spin.py

```diff
@@ -0,0 +1,504 @@
+                    self.timer.stop("train_step_time")
+                    train_step_time = self.timer.get("train_step_time")
+                    # to help avoid fragmentation
+                    clear_memory()
+
+                    # TODO(geshen): maybe use the dataloader instead
+                    # bump up the consumed samples but not the step
+                    self.consumed_samples += self.model.cfg.global_batch_size
+                    metrics["consumed_samples"] = self.consumed_samples
+                    metrics["step_time"] = train_step_time
+                    metrics["epoch"] = self.epoch
+                    metrics["iteration"] = self.iteration
+                    self.logger.log_metrics(
+                        metrics, step=self.step, prefix="train/",
+                    )
+                    metrics = {f"train_{k}": v for k, v in metrics.items()}
+
+                    self.step += 1
+
+                    run_time_exceeded = self.run_timer.is_finished()
+                    run_val, save_model, is_train_end = check_progress(
+                        self.step,
+                        self.max_steps,
+                        self.cfg.val_check_interval,
+                        self.cfg.save_interval,
+                        self.limit_val_batches,
+                        run_time_exceeded=run_time_exceeded,
+                    )
+
+                    if run_val:
+                        val_loss, val_metrics = self.run_validation()
+                        # validation is done on the UPDATED weights
+                        # so we use the incremented self.step
+                        self.logger.log_metrics(val_metrics, step=self.step, prefix="val/")
+                        val_metrics = {f"val_{k}": v for k, v in val_metrics.items()}
+                        metrics.update(val_metrics)
+
+                    global_pbar.set_postfix(metrics)
+
+                    if save_model:
+                        # PTL save wants tensors only
+                        metrics = {k: torch.as_tensor(v) for k, v in metrics.items()}
+                        self.save(metrics, is_train_end=is_train_end)
+
+                    if run_time_exceeded:
+                        logging.info(f"Time limit given by run_timer={self.run_timer} reached. Stopping run")
+                        return
+
+                    metrics.clear()
+                    self.model.finish_training()
+
+            # update the reference policy weights
+            self.model.ref_policy_state_dict = retrieve_model_state_dict_in_cpu(
+                self.model, megatron_amp_O2=self.model.cfg.get("megatron_amp_O2", False)
+            )
+
+        self.logger.finalize()
+
+    def save(self, extra_candidates=None, is_train_end=False):
+        self.model.prepare_for_training()
+        # load back in the adam states if needed
+        torch.cuda.synchronize()
+        torch.distributed.barrier()
```

Comment:
prepare for training loads back the training states from CPU -> GPU(i.e the adam states) so PTL save can save the adam optimizer. I think synchronize is probably not needed but added just to be sure all the hosttodevices are done.

In theory though, we don't need to call this to get the adam states back since they are all on CPU, but it crashes if it's not on GPU

Reply:
I see, thanks. I guess just moving the comment one line above would already help make it clearer!

Reply:
Done